### ISSUE TITLE
instID wrong in scenes with instances and user-geometry

### ISSUE THREAD

**Issue by pborsutzki** [embree/embree#75]
Hi,

I am currently trying to change our embree implementation to leverage embrees instancing api. As we also need user geometry in our renderer we have a mixed embree scene consisting of instances that refer to triangle meshes and user geometry on the same level:
```
Embree Scene (id = 0)
|-> Instance (geomID = 0)
|   |-> Embree Scene (id = 1)
|   |   |-> Triangle Mesh (geomID = 0)
|   |   |-> Triangle Mesh (geomID = 1)
|   |   |-> ...
|
|-> Instance (geomID = 1)
|   |-> EmbreeScene (id = 2)
|   |   |-> Triangle Mesh (geomID = 0)
|
|-> Instance (geomID = 1)
|   |-> Embree Scene (id = 3)
|   |   |-> Triangle Mesh (geomID = 0)
|   |   |-> User Geometry (geomID = 1)
|   |   |   |-> Custom Intersection Code
|
|-> User Geometry (geomID = 2)
|   |-> Custom Intersection Code
```
The issue I now experience: An intersection with this bvh that returns the user geometry program on the top level (`geomID = 2` in `sceneid = 0`) as closest hit gives me an `instID` that is not equal to `RTC_INVALID_GEOMETRY_ID` even though it was set to that value during initialization of the ray struct.

This becomes a problem as I am now unable to differentiate between a user geometry program that is part of an instanced scene and a user geometry program on the top level. The workaround for the time being is clear - don't mix instances with user geometry.

I sadly did not find anything on this behavior previous to implementing this in the readme. Only after noticing the problem I grepped on the source and found one comment: `[...] the instID field of a hit instance is not invalidated by other hit geometry`. So am I am missing some implementation detail of my custom intersection program (none of them touches the `instID` field at all) or is this a bug in embree? In any case if this is the expected behavior it should be documented or restricted.

Ah, and we are still using Embree 2.7.1 - just in case something has changed recently in this area that was not part of the changelog.

Best Regards,
Philipp

**Comment by svenwoop** [comment]
We handle instances the following way internally. All standard geometries (triangles, quads, etc.) are intersected first, and do not modify the instID field of the ray. After that the user geometries are intersected, and instances are just a user geometry that we create internally.

What happens in your case is that you hit an instance which sets the instID field of the ray. Then you hit your user geometry, which likely does not set the instID of the ray. That is an issue.

When you use instances the instID is a hit field that is required to be set. Thus your user geometries have to set the instID field (e.g. to RTC_INVALID_GEOMETRY).

I will check if that behavior is documented.

**Comment by pborsutzki** [comment]
That means my user geometries that are _not instanced_ must know that they are not instanced and thus need to reset the `instID` field while my _instanced user geometries_ must know that they are instanced and must not reset the `instID` field? Sounds a bit inconvenient but I probably can implement such behavior.

Thank you for the very fast help!

**Comment by svenwoop** [comment]
Yes unfortunately that is true. Alternatively you can also instantiate all geometries.
